qubes/vm/dispvm.py

```diff
@@ -187,11 +187,37 @@ def __init__(self, app, xml, *args, **kwargs):
             self.features.update(template.features)
             self.tags.update(template.tags)
 
+    def is_domain_preloaded(self):
+        preload_dispvm = (
+            self.features.check_with_template("preload-dispvm", None)
+        )
+        if not preload_dispvm:
+            return False
+        preload_dispvm = preload_dispvm.split(" ")
+        if self.name not in preload_dispvm:
+            return False
+        return True
+
+
     @qubes.events.handler("domain-load")
     def on_domain_loaded(self, event):
         """When domain is loaded assert that this vm has a template."""  # pylint: disable=unused-argument
         assert self.template
 
+    @qubes.events.handler("domain-started")
+    def on_domain_started(self, event, **kwargs):
+        """Pause preloaded domains as soon as they start."""
+        if self.is_domain_preloaded():
+            self.pause()
+
+    @qubes.events.handler("domain-unpaused")
+    def on_domain_unpaused(self):
+        """Mark unpaused preloaded domains as used."""
+        if self.is_domain_preloaded():
+            self.app.qubesd_call(
```

Comment:
This doesn't exist in core-admin - it's only in core-admin-client. Here, you are running inside qubesd already, you can simply call appropriate function directly.